src/armscan_env/volumes/slicing.py

```diff
@@ -0,0 +1,238 @@
+        )
+
+        return transformed_action
+
+
+class TransformedVolume(sitk.Image):
+    """Represents a volume that has been transformed by an action.
+
+    Should only ever be instantiated by `create_transformed_volume`.
+    """
+
+    def __init__(self, *args, transformation_action: ManipulatorAction, _private: int):
+        if _private != 42:
+            raise ValueError(
+                "TransformedVolume should only be instantiated by create_transformed_volume.",
+            )
+        super().__init__(*args)
+        self._transformation_action = transformation_action
+
+    @property
+    def transformation_action(self) -> ManipulatorAction | None:
+        return self._transformation_action
+
+
+def create_transformed_volume(
+    volume: sitk.Image,
+    transformation_action: ManipulatorAction,
+) -> TransformedVolume:
+    """Transform a 3D volume with arbitrary rotation and translation.
+
+    :param volume: 3D volume to be transformed
+    :param transformation_action: action to transform the volume
+    :return: the sliced volume.
+    """
+    if isinstance(volume, TransformedVolume):
+        raise ValueError(
+            f"This operation should only be performed on a non-transformed volume "
+            f"but got an instance of: {volume.__class__.__name__}.",
+        )
+
+    origin = np.array(volume.GetOrigin())
+    transform_matrix = EulerTransform(transformation_action, origin).get_transform_matrix()
+
+    # Define plane's coordinate system
+    e1 = transform_matrix[0][:3]
+    e2 = transform_matrix[1][:3]
+    e3 = transform_matrix[2][:3]
+    img_o = transform_matrix[:, -1:].flatten()[:3]  # origin of the image plane
+
+    direction = np.stack([e1, e2, e3], axis=0).flatten()
+
+    resampler = sitk.ResampleImageFilter()
+    spacing = volume.GetSpacing()
+
+    resampler.SetOutputDirection(direction.tolist())
+    resampler.SetOutputOrigin(img_o.tolist())
+    resampler.SetOutputSpacing(spacing)
+    resampler.SetSize(volume.GetSize())
+    resampler.SetInterpolator(sitk.sitkNearestNeighbor)
+
+    # Resample the volume on the arbitrary plane
+    transformed_volume = resampler.Execute(volume)
+    # needed to deal with rotation dependency of the volume
+    return TransformedVolume(
+        transformed_volume,
+        transformation_action=transformation_action,
+        _private=42,
+    )
+
+
+def get_volume_slice(
+    volume: sitk.Image,
+    # TODO: shouldn't there be a default shape, like the native shape of the volume itself?
+    slice_shape: tuple[int, int],
+    action: ManipulatorAction,
+) -> sitk.Image:
+    """Slice a 3D volume with arbitrary rotation and translation.
+
+    :param volume: 3D volume to be sliced
+    :param slice_shape: shape of the output slice
+    :param action: action to transform the volume
+    :return: the sliced volume.
+    """
+    o = np.array(volume.GetOrigin())
+
+    if isinstance(volume, TransformedVolume):
+        # TODO: why is origin not used here?
```

Comment:
@carlocagnetta pls address the TODOs in this function :)

Reply:
this whole thing is a bit of a brain teaser actually.